akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala

```diff
@@ -90,4 +96,91 @@ object StreamConverters {
   def asInputStream(readTimeout: FiniteDuration = 5.seconds): Sink[ByteString, InputStream] =
     Sink.fromGraph(new InputStreamSinkStage(readTimeout))
 
+  /**
+    * A `Sink` that will use Java 8 `Collector` to reduce and optionally transform
+    * accumulated result into a final representation.
+    * As collector is mutable you can use it only with one stream.
+    * Reduction operations perform sequentially.
+    */
+  def javaCollector[T, R](collectorFactory: () ⇒ java.util.stream.Collector[T, _ <: Any, R]): Sink[T, Future[R]] =
+    Flow[T].fold(() ⇒
+      new CollectorState[T,R](collectorFactory().asInstanceOf[Collector[T, Any, R]])) { (state, elem) ⇒ () ⇒ state().update(elem) }
+      .map(state ⇒ state().finish())
+      .toMat(Sink.head)(Keep.right).named("javaCollectorSink")
+
+  /**
+    * A `Sink` that will use Java 8 `Collector` to reduce and optionally transform
+    * accumulated result into a final representation.
+    * As collector is mutable you can use it only with one stream.
+    * Reduction operations perform in parallel.
+    */
+  def javaCollectorParallelUnordered[T, R](parallelism: Int)(collectorFactory: () ⇒ java.util.stream.Collector[T, _ <: Any, R]): Sink[T, Future[R]] = {
+    if (parallelism == 1) javaCollector[T, R](collectorFactory)
+    else {
+      val createOnceFactory = () => {
+        val ref:AtomicReference[Collector[T, Any, R]] = new AtomicReference[Collector[T, Any, R]]()
+        ref.compareAndSet(null, collectorFactory().asInstanceOf[Collector[T, Any, R]])
+        ref.get()
+      }
```

Comment:
Why is this needed? Couldn't the GraphDSL part just call `collectorFactory()` once and then use that instance to each `CollectorState` it creates?